packages/data/src/plugins/actions.js

```diff
@@ -125,6 +126,62 @@ export function* installAndActivatePlugins( plugins ) {
 	}
 }
 
+export const createErrorNotice = ( errorMessage ) => {
+	return dispatch( 'core/notices', 'createNotice', errorMessage );
+};
+
+export function* installJetpackAndConnect() {
+	try {
+		yield dispatch( STORE_NAME, 'installPlugins', [ 'jetpack' ] );
+		yield dispatch( STORE_NAME, 'activatePlugins', [ 'jetpack' ] );
+
+		const url = yield select( STORE_NAME, 'getJetpackConnectUrl', {
+			redirect_url: getAdminLink( 'admin.php?page=wc-admin' ),
+		} );
+
+		// getJetpackConnectUrl doesn't throw errors, it sets error state if
+		// something went wrong.
+		const error = yield select(
+			STORE_NAME,
+			'getPluginsError',
+			'getJetpackConnectUrl'
+		);
+
+		if ( error ) {
+			yield createErrorNotice( error );
+		} else {
+			window.location = url;
+		}
+	} catch ( error ) {
+		yield createErrorNotice( error.message );
+	}
+}
+
+export function* connectToJetpack( failureRedirect ) {
+	try {
+		const url = yield select( STORE_NAME, 'getJetpackConnectUrl', {
+			redirect_url: getAdminLink( 'admin.php?page=wc-admin' ),
+		} );
+
+		// getJetpackConnectUrl doesn't throw errors, it sets error state if
+		// something went wrong.
+		const error = yield select(
+			STORE_NAME,
+			'getPluginsError',
+			'getJetpackConnectUrl'
+		);
+
+		if ( error ) {
+			yield createErrorNotice( error );
+			window.location = failureRedirect;
+		} else {
+			window.location = url;
```

Comment:
I have slightly mixed feelings about setting window location from a data store.  It feels like a decision that a data store should not be making.  That said, I think this action has a very obvious goal and consumers will be aware of this.

Reply:
@joshuatf Personally I think its ok for the reason that this is how I'd tackle this in redux saga as well. usually it would be some kind of yield of a client side routing call, but basically redux-saga and our generator actions want you to do side effect stuff here versus inside components. I don't really like the pattern we currently have of trying to render a component and then if some particular state is met mutating window.location there instead. That said some of that distaste of redirecting inside a component may be my own personal bias.